File: 0300-longest-increasing-subsequence/solution.py
class Solution(object):
    def lengthOfLIS(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        
        dp = []
        
        for i in range(len(nums)):
            idx = bisect_left(dp, nums[i])
            
            if idx == len(dp):
                dp.append(nums[i])
            else:
                dp[idx] = nums[i]
        
        return len(dp)

File: 0300-longest-increasing-subsequence/solution.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) 
    {
        vector<int> subSequence;
        
        for(int i = 0; i < nums.size(); i++)
        {
            auto itr = lower_bound(subSequence.begin(), subSequence.end(), nums[i]);
            if(itr == subSequence.end())
            {
                subSequence.push_back(nums[i]);
            }
            else
            {
                *itr = nums[i];
            }
        }
        
        return subSequence.size();
        
    }
};
